### DISCUSSION TITLE
Filtering Detected Objects in Ultralytics YOLOv8 Based on Class and Confidence Threshold

### DISCUSSION THREAD

**Discussion by aiyafi** [ultralytics/ultralytics#13354] [General]
Currently, I'm a student who working on a human detection project using Ultralytics YOLOv8 or “Loitering Detection in an Intelligent Surveillance System using YOLO-based Methods”. I want to display detections only when the object is classified as a "person" and has a confidence score above 0.8.

I've added the following line to set the confidence threshold:
```confidence_threshold = 0.8  # Set custom threshold for the 'person' class```

However, the code doesn't seem to filter the results as expected.  Detections with lower confidence scores or incorrect classifications are still being displayed. Just like this screenshot.

![image](https://github.com/ultralytics/ultralytics/assets/63116789/c4e084af-7e60-428b-8dff-8ae4305ce7e6)

and here's my code:

```
from ultralytics import YOLO
import cv2
import csv
import random

# --- Random Character Generation ---
def generate_random_char():
    random_number = random.randint(0, 99)
    return str(random_number).zfill(2)

confidence_threshold = 0.8  # Set custom threshold for the 'person' class

# load yolov8 model
model = YOLO('yolov8n.pt')

# --- Naming Setup ---
random_char = generate_random_char() # Generate random 2-digit character
# csv_file_path = 'centroid_data.csv'
csv_file_path = f'centroid_data_{random_char}.csv'
with open(csv_file_path, 'w', newline='') as csvfile:
    csv_writer = csv.writer(csvfile)
    csv_writer.writerow(['x', 'y'])  # Write the header row
    
video_file_path = f'video_{random_char}.mp4'  # Naming the video file
# -----

# load video
video_path = 'D:\\@Dev\\finalPrj\\objectDetection1\\480%Pgm-Abnormal Cowok Pagi.mp4'
cap = cv2.VideoCapture(video_path)

# Get video properties for output
width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
fps = int(cap.get(cv2.CAP_PROP_FPS))

print(f"Video Width: {width}, Video Height: {height}")

# Create video writer object
fourcc = cv2.VideoWriter_fourcc(*'avc1')
out = cv2.VideoWriter(video_file_path, fourcc, fps, (width, height))

# Main processing loop
while True:
    ret, frame = cap.read()

    # Write frame to video
    if not ret:
        break  # End of video

    # Object detection
    results = model.predict(frame, classes=[0], save=True, save_txt=True)

    # --- Save Centroid to CSV ---
    for r in results:  
        for i, box in enumerate(r.boxes.xyxy):
            centroid_y = (box[1] + box[3]) / 2  # Use YOLO's y1 and y2 
            centroid_x = (box[0] + box[2]) / 2  # Use YOLO's x1 and x2

            # Draw bounding box and add label + confidence score
            label = f'{r.names[int(r.boxes.cls[i])]} {r.boxes.conf[i].item():.2f}'
            cv2.rectangle(frame, (int(box[0]), int(box[1])), (int(box[2]), int(box[3])), (0, 255, 0), 2)

            # Print and/or draw on the frame
            print(f"Centroid Coordinates: ({centroid_x}, {centroid_y})\n")
            print(box, "\n")  # Print YOLO bounding box coordinates
            cv2.circle(frame, (int(centroid_x), int(centroid_y)), radius=2, color=(0, 0, 255), thickness=3) # Draw centroid
            
            # Save to CSV
            with open(csv_file_path, 'a', newline='') as csvfile:  # 'a' for append
                csv_writer = csv.writer(csvfile)
                csv_writer.writerow([centroid_x, centroid_y])

    # Write frame to video
    for result in results:
        frame_ = result.plot()  # Get frame with bounding boxes
        out.write(frame_)
        cv2.imshow('OBJ Detection', frame_)  # Display frame

    # Exit condition
    if cv2.waitKey(25) & 0xFF == ord('q'):
        break

# Release resources
cap.release()
out.release()  # Release the video writer
cv2.destroyAllWindows()
```

Any suggestions on how to correctly filter the detections based on both class and confidence in YOLOv8?

**Comment by pderrenger** [comment]
Hello! It looks like you're on the right track with setting up your YOLOv8 model for person detection. However, to ensure that only detections of the "person" class with a confidence score above 0.8 are displayed, you'll need to adjust how you filter these results in your code.

Here’s a quick adjustment you can make:

1. **Set the confidence threshold directly in the `predict()` method** rather than setting it as a separate variable. This ensures that the model uses this threshold during prediction.
2. **Specify the class ID for 'person'** in the `classes` parameter if you haven't already. The class ID for 'person' is typically `0` in most pre-trained models like COCO.

Here's how you can modify your prediction line:

```python
results = model.predict(frame, conf=0.8, classes=[0])
```

This line will ensure that the model only returns detections for the "person" class with a confidence score of at least 0.8.

If you continue to see unexpected results, make sure that the model is loaded correctly and that the class IDs correspond correctly to those in your model's training dataset.

Let me know if this helps or if you have any more questions!

**Reply by aiyafi** [reply]
Hello [pderrenger](https://github.com/pderrenger)! Just wanted to say a huge thank you for your help. Your insights were incredibly valuable and made a real difference in my understanding of the topic. I really appreciate you taking the time to share your knowledge.

**Reply by aiyafi** [reply]
Hey, thx for the help :)

**Reply by pderrenger** [reply]
You're welcome! I'm glad I could assist. If you have any more questions or need further help in the future, feel free to reach out. Happy coding! 😊
